## Introduction ##
Feature selection is a pattern recognition approach to choose important variables according to some criteria in order to distinguish or explain certain phenomena (i.e., for dimensionality reduction). There are many genomic and proteomic applications that rely on feature selection to answer questions such as selecting signature genes which are informative about some biological state, e.g., normal tissues and several types of cancer; or inferring a prediction network among elements such as genes, proteins and external stimuli. In these applications, a recurrent problem is the lack of samples to perform an adequate estimate of the joint probabilities between element states. A myriad of feature selection algorithms and criterion functions have been proposed, although it is difficult to point the best solution for each application.

The intent of this software is to provide an open-source multiplataform graphical environment for bioinformatics problems, which supports many feature selection algorithms, criterion functions and graphic visualization tools such as scatterplots, parallel coordinates and graphs. A feature selection approach for growing genetic networks from seed genes (targets or predictors) is also implemented in the system.

The feature selection environment allows data analysis using several algorithms, criterion functions and graphic visualization tools. Our experiments have shown the software effectiveness in two distinct types of biological problems. Besides, the environment can be used in different pattern recognition applications, although the main concern regards bioinformatics tasks.


---


## Videos ##

Video demonstration of the DimReduction GUI use to the feature selection and classification approach.
http://www.youtube.com/watch?v=CZO0t0l1kwk

Video demonstration of the DimReduction GUI use to the feature selection approach for growing genetic networks from targets or predictors genes.
http://www.youtube.com/watch?v=h8GWn5XbOGo


---


## Main References: ##

F. M. Lopes, D. C. Martins-Jr and R. M. Cesar-Jr. **Feature selection environment for genomic applications**, BMC Bioinformatics, 9 (451), 2008.
http://dx.doi.org/10.1186/1471-2105-9-451

F. M. Lopes, D. C. Martins-Jr and R. M. Cesar-Jr. **DimReduction - Interactive Graphic Environment for Dimensionality Reduction**.  Technical Report, 2008.
http://arxiv.org/abs/0805.3964v1


## Additional References: ##
F. F. R. Vicente, F. M. Lopes, R. F. Hashimoto, R. M. Cesar-Jr. **Assessing the Gain of Biological Data Integration in Gene Networks Inference**. BMC Genomics, v. 13, p. S7, 2012.
http://dx.doi.org/10.1186/1471-2164-13-S6-S7

F. M. Lopes, R. M. Cesar-Jr and L. da F. Costa. **Gene expression complex networks: synthesis, identification and analysis**. Journal of Computational Biology, v. 18, p. 1353-1367, 2011.
http://dx.doi.org/10.1089/cmb.2010.0118

F. M. Lopes, E. A. de Oliveira and R. M. Cesar-Jr. **Inference of gene regulatory networks from time series by Tsallis entropy**, BMC Systems Biology, 5 (61), 2011.
http://dx.doi.org/10.1186/1752-0509-5-61

F. M. Lopes, D. C. Martins-Jr, J. Barrera and R. M. Cesar-Jr. **SFFS-MR: a floating search strategy for GRNs inference**.  In: 5th IAPR International Conference on Pattern Recognition in Bioinformatics - PRIB, 2010, Nijmegen, The Netherlands.
http://dx.doi.org/10.1007/978-3-642-16001-1_35

F. M. Lopes, E. A. de Oliveira and R. M. Cesar-Jr. **Analysis of the GRNs inference by using Tsallis entropy and a feature selection approach**.  In: 14th Iberoamerican Congress on Pattern Recognition - CIARP, 2009, Guadalajara, JA, Mexico.
http://dx.doi.org/10.1007/978-3-642-10268-4_55

F. M. Lopes, D. C. Martins-Jr and R. M. Cesar-Jr. **Comparative study of GRNs inference methods based on feature selection by mutual information**.  In: 7th IEEE International Workshop on Genomic Signal Processing and Statistics - GENSIPS, 2009, Minneapolis, MN, USA.
http://dx.doi.org/10.1109/GENSIPS.2009.5174334

F. M. Lopes, R. M. Cesar-Jr and L. da F. Costa. **AGN simulation and validation model**. In Advances in Bioinformatics and Computational Biology, Proceedings, volume 5167 of Lecture Notes in Bioinformatics, pages 169â€“173, 2008. Springer-Verlag Berlin.
http://dx.doi.org/10.1007/978-3-540-85557-6_17


---


## Other works developed by the Creativision group and Bioinfo-CP: ##

**jagn: A Java-Based Model for Artificial Gene Networks Generation**
http://code.google.com/p/jagn/

**Creative Vision Research Group**
http://www.vision.ime.usp.br/~creativision/

**Bioinfo-CP: Bioinformatics and Pattern Recognition Research Group**
http://www.utfpr.edu.br/cornelioprocopio/pesquisa/grupos-de-pesquisa/bioinfo-cp/about-us/about-us