<html>    <head>        <title>Help - Single Execution Panel / Network Identification</title>    </head>    <body>        <br>        <p style="margin-top: 0">            <font face="Tahoma, Arial" size="5"><b>This panel is dedicated to perform single execution of feature selection and graph generation.</b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>Implemented Functionalities:</b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Combo Box) Criterion Function:</b> Select the criterion function based on classifier information (mean conditional entropy) or based on classifier error (CoD - Coefficient of Determination).</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Field) q-entropy:</b> The value defined in this text field represents the type of entropy considered in criterion function. Use 1 to apply Shannon Entropy. Use a value different from 1 to apply Tsallis Entropy (smaller than 1 for subextensive entropy or larger than 1 for superextensive entropy). This parameter is used only by the criterion function based on entropy.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Combo Box) Penalization Method:</b></font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- no_obs:</b> apply penalty for non-observed instances.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- poor_obs:</b> apply penalty for poorly observed instances (observed only once).</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Alpha (value for penalty):</b> Only used by no_obs penalization method. Alpha value represents the probability mass for the non-observed instances. This parameter is added to the relative frequency (number of occurrences) of all possible instances.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Slider Bar) Beta (value of confidence):</b> Only used by poor_obs penalization method. Beta value is attributed to the conditional probability of the observed class, given the instance. The value (1-Beta) is equally distributed over non-observed classes of the considered instance.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Input Test Set (optional):</b> Click on File button or fill this text box with the path of a data file, which is used to perform test of the classifier. If this text field has no information, the input data file is used to perform training and test of the classifier (resubstitution error method).</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) File:</b> Shows open dialog to select test data file. The path of the selected file is stored automatically at Input Test Set text field.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Execute Feature Selector:</b> Execute feature selection on input data (if input data set is quantized) or on quantized data. The data must have the label of the classes in its last column.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Progress Bar):</b> Display the progress of the feature selection execution.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Option Group) Search Method:</b></font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- SFS:</b> Apply SFS algorithm to perform feature selection.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- SFFS:</b> Apply SFFS algorithm to perform feature selection.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- Exhaustive Search:</b> Apply Exhaustive algorithm for feature selection.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Maximum Set Size:</b> Select the maximum cardinality of a feature set to perform the search. Use this option only for SFFS algorithm.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Size of the Result List:</b> This parameter allows to choose the size of the result list with the first ranked features sets.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Check Box) Targets as Predictors?:</b> Select this option to generate graph from targets (not selected) or predictors (selected).</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Target's indexes:</b> Fill this text box with predictors or targets indexes to find others variables (genes) related with them, and then click in the Network Inference button. If this text box is empty, all features are considered to graph generation.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Option Group):</b></font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- time-series data:</b> Select this option if your data represents a time-series gene expressions, i.e., it is considered by the method time-dependent relationship among the variables/genes (rows) and its observations/samples (columns). The predictors are observed at time t and the targets are observed at time t+1.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- steady-state data:</b> Select this option if your data represents an independent gene expressions, i.e., it is considered by the method relationships among variables/genes (rows) only within each experiment (columns).</font></p        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Check Box) Is it periodic?:</b> Mark this option to assume that time series is periodic, i. e. the last instant of time is connected to the first instant of time.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Field) Threshold:</b> Fill this text box with a real value to visualize all graph edges (near 1) or just most representative ones (near 0).</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Network Inference:</b> Apply feature selection algorithm to find relationship among variables (genes), and as a result, display a network in which variables are presented as nodes and its relationships as edges.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Scatter Plot:</b> Display a scatter plot chart with two selected features indexes (columns) of the input data.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Parallel Coordinates:</b> Display parallel coordinates chart with selected features indexes (columns) from the input data.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Check Box) Average values only:</b> Mark this check box to plot only the average values of the features (columns) in parallel coordinates chart.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Check Box) Normalize values:</b> Mark this check box to plot normalized values between 0 and 1 of the features (columns) in parallel coordinates chart.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Signal Plot:</b> Display a signal plot of the selected variables indexes (rows) of the input data.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Correlation Coefficient:</b> Display a table with correlation coefficient of the selected features indexes (columns) of the input data.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) View Inferred Network:</b> Displays a graph in which variables are pesented as nodes and its relationships as edges.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Text Area) Results:</b></font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- Classification results (right side):</b> Shows the criterion value for the selected features.</font></p>        <p style="margin-top: 0; margin-left: 1cm"><font face="Tahoma, Arial" size="4"><b>- Feature Selection results (left side):</b> Shows targets and its predictors or predictors and its targets, and the criterion value. If ties occur, all the tied sets are displayed.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>- (Button) Save Results:</b> Click to save results into a text file.</font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>Web page of DimReduction project are available at <a href="http://dimreduction.incubadora.fapesp.br">http://dimreduction.incubadora.fapesp.br</a>.</b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>Source code, binaries, javadoc and videos demonstrations of DimReduction are available at <a href="http://dimreduction.incubadora.fapesp.br">http://dimreduction.incubadora.fapesp.br</a> and <a href="http://sourceforge.net/projects/dimreduction">http://sourceforge.net/projects/dimreduction</a>.</b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b><i>References:</i></b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>                    F. M. Lopes, D. C. Martins-Jr and R. M. Cesar-Jr. <a href="http://www.biomedcentral.com/1471-2105/9/451" target="_blank">Feature selection environment for genomic applications</a>, BMC Bioinformatics, 9(451), 2008.        </b></font></p>        <br>        <p style="margin-top: 0"><font face="Tahoma, Arial" size="4"><b>                    A detailed technical report are available at <a href="http://arxiv.org/abs/0805.3964">http://arxiv.org/abs/0805.3964</a>.        </b></font></p>        <br>    </body></html>